src/ray/gcs/gcs_server/gcs_resource_manager.cc

```diff
@@ -341,5 +365,110 @@ std::string GcsResourceManager::ToString() const {
   return ostr.str();
 }
 
+void GcsResourceManager::HandleGetClusterResourceState(
+    rpc::autoscaler::GetClusterResourceStateRequest request,
+    rpc::autoscaler::GetClusterResourceStateReply *reply,
+    rpc::SendReplyCallback send_reply_callback) {
+  // TODO: last_seen_cluster_resource_state_version not needed?
+  RAY_CHECK(request.last_seen_cluster_resource_state_version() <=
+            cluster_resource_state_version_);
+
+  reply->set_last_seen_autoscaler_state_version(last_seen_autoscaler_state_version_);
+  reply->set_cluster_resource_state_version(cluster_resource_state_version_);
+
+  GetNodeStates(reply);
+  GetPendingResourceRequests(reply);
+  GetPendingGangResourceRequests(reply);
+  GetClusterResourceConstraints(reply);
+
+  // QQ: should we use GCS_RPC_SEND_REPLY
+  // That would add another field in the reply, and introduce dep of `GcsStatus`.
+  send_reply_callback(ray::Status::OK(), nullptr, nullptr);
+}
+
+void GcsResourceManager::GetPendingGangResourceRequests(
+    rpc::autoscaler::GetClusterResourceStateReply *reply) {
+  // TODO
+  return;
+}
+
+void GcsResourceManager::GetClusterResourceConstraints(
+    rpc::autoscaler::GetClusterResourceStateReply *reply) {
+  // TODO
+  return;
+}
+
+void GcsResourceManager::GetPendingResourceRequests(
+    rpc::autoscaler::GetClusterResourceStateReply *reply) {
+  std::unordered_map<google::protobuf::Map<std::string, double>, rpc::ResourceDemand>
+      aggregate_load;
+
+  for (const auto &usage : node_resource_usages_) {
+    // Aggregate the load reported by each raylet.
+    FillAggregateLoad(usage.second, &aggregate_load);
+  }
+
+  for (const auto &[shape, demand] : aggregate_load) {
+    // QQ: should backlog size to be included here?
+    auto num_pending = demand.num_infeasible_requests_queued();
```

Comment:
Hmm pending != infeasible?

Reply:
should we include backlog here?

Reply:
hmm i think from autoscaler's perspective it can differentiate pending from infeasible, so we don't need to differentiate them. @rkooo567 does it make sense?

but from core's perspective pending is different from infeasible.